Libplanet.Net/Consensus/Context.Async.cs

```diff
@@ -0,0 +1,189 @@
+                    throw;
+                }
+            }
+        }
+
+        private async Task ConsumeMessage(CancellationToken cancellationToken)
+        {
+            ConsensusMessage message = await _messageRequests.Reader.ReadAsync(cancellationToken);
+            ProduceMutation(() =>
+            {
+                try
+                {
+                    AddMessage(message);
+                    _logger.Debug(
+                        "{FName}: Message: {Message} => " +
+                        "Height: {Height}, Round: {Round}, Address: {Address}, " +
+                        "Hash: {BlockHash}, MessageCount: {Count}. (context: {Context})",
+                        nameof(AddMessage),
+                        message,
+                        message.Height,
+                        message.Round,
+                        message.Remote!.Address,
+                        message.BlockHash,
+                        _messagesInRound[message.Round].Count,
+                        ToString());
+                }
+                catch (Exception e)
+                {
+                    _logger.Error(
+                        e,
+                        "Unexpected exception occurred during {FName}.",
+                        nameof(AddMessage));
+                }
+            });
+            ProduceMutation(() => ProcessHeightOrRoundUponRules(message));
+            MessageConsumed?.Invoke(this, message);
+        }
+
+        private async Task ConsumeMutation(CancellationToken cancellationToken)
+        {
+            System.Action mutation = await _mutationRequests.Reader.ReadAsync(cancellationToken);
+            (int MessageLogSize, int Round, Step Step) prevState =
+                (_messagesInRound.Sum(x => x.Value.Count), Round, Step);
+            mutation();
+            (int MessageLogSize, int Round, Step Step) nextState =
+                (_messagesInRound.Sum(x => x.Value.Count), Round, Step);
+            if (prevState != nextState)
+            {
+                _logger.Debug(
+                    "State (MessageLogSize, Round, Step) changed from " +
+                    "({PrevMessageLogSize}, {PrevRound}, {PrevStep}) to " +
+                    "({NextMessageLogSize}, {NextRound}, {NextStep})",
+                    prevState.MessageLogSize,
+                    prevState.Round,
+                    prevState.Step.ToString(),
+                    nextState.MessageLogSize,
+                    nextState.Round,
+                    nextState.Step.ToString());
+                StateChanged?.Invoke(
+                    this, (nextState.MessageLogSize, nextState.Round, nextState.Step));
+
+                // FIXME: This is to avoid an exception.
+                // Methods accessing message log should be changed instead.
+                if (_messagesInRound.ContainsKey(Round))
+                {
+                    ProduceMutation(() => ProcessGenericUponRules());
```

Comment:
Why produce mutation instead of just call it directly? Seems like `ConsumeMutation` is single threaded.

Reply:
Well, running `ProcessGenericUponRules()` can result in additional state changes, so no, that would possibly result in a different behavior unless we run a loop. 😶

Reply:
But isn't `ProcessGenericUponRules` just put mutation into the mutastion channel?

Reply:
I don't think I understand the question.

- `ProduceMutation()` puts a mutation, in this case `ProcessGenericUponRules()`, into the mutation channel.
- `MutationConsumerTask()` takes it out of the queue and consumes it at some point.
- `ProcessGenericUponRules()` actually mutates `Context` when certain conditions are met.

Reply:
Now that I look at it again, there is an issue with split conditional and body. 😥
At the time of actually running `ProcessGenericUponRules()` does not guarantee `_messagesInRound` contains `Round`. Should be easy enough to fix I think. 😗

Reply:
Arg... seems like this isn't that easy to fix. Probably have to deal with `FIXME` comment right above the highlighted body. 😥

Reply:
See de40c5485070e83b0df34b289da2ec9494bf86c2.